src/group_theory/divisible.lean

```diff
@@ -0,0 +1,292 @@
+* `monoid.pow_surj_of_rootable_by` : the constructive definition of rootablity implies the
+  condition that `xⁿ = y` has solutions for all `n ≠ 0` and `a ∈ A`.
+* `monoid.rootable_by_prod` : any product of two rootable monoids is rootable.
+* `monoid.rootable_by_pi` : any product of rootable monoids is rootable.
+* `group.rootable_by_int_of_rootable_by_nat` : in groups, int rootablity is implied by nat
+  rootablity.
+* `group.rootable_by_nat_of_rootable_by_int` : in groups, nat rootablity is implied by int
+  rootablity.
+* `comm_group.rootable_by_quotient` : quotient group of rootable group is rootable.
+* `comm_group.rootable_by_of_surj` : if `A` is rootable and `A →* B` is surjective, then `B` is
+  rootable.
+
+TODO: Show that divisibility implies injectivity in the category of `AddCommGroup`.
+-/
+
+open_locale pointwise
+
+namespace add_monoid
+
+variables (A α : Type*) [add_monoid A] [has_smul α A] [has_zero α]
+
+/--
+An `add_monoid A` is `α`-divisible iff `n • x = a` has a solution for all `n ≠ 0 ∈ α` and `a ∈ A`.
+Here we adopt a constructive approach where we ask an explicit `div : A → α → A` function such that
+* `div a 0 = 0` for all `a ∈ A`
+* `n • div a n = a` for all `n ≠ 0 ∈ α` and `a ∈ A`.
+-/
+class divisible_by :=
+(div : A → α → A)
+(div_zero : ∀ a, div a 0 = 0)
+(div_cancel : ∀ {n : α} (a : A), n ≠ 0 → n • (div a n) = a)
+
+section pi
+
+variables {ι β : Type*} (B : ι → Type*) [Π (i : ι), has_smul β (B i)]
+variables [has_zero β] [Π (i : ι), add_monoid (B i)] [Π i, divisible_by (B i) β]
+
+instance divisible_by_pi : divisible_by (Π i, B i) β :=
+{ div := λ x n i, divisible_by.div (x i) n,
+  div_zero := λ x, funext $ λ i, divisible_by.div_zero _,
+  div_cancel := λ n x hn, funext $ λ i, divisible_by.div_cancel _ hn }
+
+end pi
+
+section prod
+
+variables {β B B' : Type*} [has_zero β] [add_monoid B] [add_monoid B']
+variables [has_smul β B] [has_smul β B'] [divisible_by B β] [divisible_by B' β]
+
+instance divisible_by_prod : divisible_by (B × B') β :=
+{ div := λ p n, (divisible_by.div p.1 n, divisible_by.div p.2 n),
+  div_zero := λ p, prod.ext (divisible_by.div_zero _) (divisible_by.div_zero _),
+  div_cancel := λ n p hn, prod.ext (divisible_by.div_cancel _ hn) (divisible_by.div_cancel _ hn) }
+
+end prod
+
+end add_monoid
+
+namespace monoid
+
+variables (A α : Type*) [monoid A] [has_pow A α] [has_zero α]
+
+/--
+A `monoid A` is `α`-rootable iff `xⁿ = a` has a solution for all `n ≠ 0 ∈ α` and `a ∈ A`.
+Here we adopt a constructive approach where we ask an explicit `root : A → α → A` function such that
+* `root a 0 = 1` for all `a ∈ A`
+* `(root a n)ⁿ = a` for all `n ≠ 0 ∈ α` and `a ∈ A`.
+-/
+@[to_additive add_monoid.divisible_by
+"An `add_monoid A` is `α`-divisible iff `n • x = a` has a solution for all `n ≠ 0 ∈ α` and `a ∈ A`.
+Here we adopt a constructive approach where we ask an explicit `div : A → α → A` function such that
+* `div a 0 = 0` for all `a ∈ A`
+* `n • div a n = a` for all `n ≠ 0 ∈ α` and `a ∈ A`."]
```

Comment:
There's no need to repeat the docstring here, since you already have it above.

Reply:
removed